packages/web3-wallet/src/hd-wallet.ts

```diff
@@ -60,3 +67,67 @@ export function getHDWalletPath(addressIndex: number): string {
 
   return `m/44'/${coinType}/0'/0/${addressIndex}`
 }
+
+export type HDWalletAccount = Account & { addressIndex: number }
+
+export class HDWallet extends SignerProviderWithCachedAccounts<HDWalletAccount> {
+  private readonly mnemonic: string
+  private readonly passphrase?: string
+  readonly nodeProvider: NodeProvider
+  readonly explorerProvider: ExplorerProvider | undefined
+
+  constructor(mnemonic: string, nodeProvider?: NodeProvider, explorerProvider?: ExplorerProvider, passphrase?: string) {
+    super()
+    this.mnemonic = mnemonic
+    this.passphrase = passphrase
+    this.nodeProvider = nodeProvider ?? web3.getCurrentNodeProvider()
+    this.explorerProvider = explorerProvider ?? web3.getCurrentExplorerProvider()
+  }
+
+  private getNextFromAddressIndex(targetGroup?: number): number {
+    let usedAddressIndex = -1
+    for (const account of this._accounts.values()) {
+      if ((targetGroup === undefined || account.group == targetGroup) && account.addressIndex > usedAddressIndex) {
+        usedAddressIndex = account.addressIndex
+      }
+    }
+    return usedAddressIndex + 1
+  }
+
+  deriveNewAccount(targetGroup?: number): HDWalletAccount {
```

Comment:
Q: When we derive a new account, we might have already generated/persisted a few accounts from before. Would it be good to allow passing in the `addressIndex`es that are already used?
IIUC, the cached accounts here is not persisted, right?

Reply:
Yes, it's done in this line:
```
    const fromAddressIndex = this.getNextFromAddressIndex(targetGroup)
```
The derivation will start from the smallest available index.

Reply:
Yes, that's totally right. This is an in-memory wallet designed for non-critical use cases, for example, contract deployment. For productions wallet (e.g. extension wallet), those derivation functions should be used instead. This wallet is not generic or flexible enough for production wallets. Same for `PrivateKeyWallet`. I will add a comment on this.

WDYT?

Reply:
Btw, the derived addresses are added to `_accounts` if it's derived with `deriveAndAddNewAccount`.

Reply:
Got it, sounds great :+1:

Reply:
Just simplified HDWallet to remove `deriveNewAccount` and keep `deriveAndAddNewAccount` only.